subsys/bluetooth/services/ots_olcp.c

```diff
@@ -0,0 +1,311 @@
+	struct bt_gatt_ots *ots)
+{
+	int err;
+	struct bt_gatt_ots_object *prev_obj;
+
+	if (!ots->cur_obj) {
+		return BT_GATT_OTS_OLCP_RES_OPERATION_FAILED;
+	}
+	err = bt_gatt_ots_obj_manager_prev_obj_get(ots->obj_manager,
+						   ots->cur_obj,
+						   &prev_obj);
+	if (err) {
+		return obj_manager_to_olcp_err_code_map(err);
+	}
+
+	ots->cur_obj = prev_obj;	
+
+	return BT_GATT_OTS_OLCP_RES_SUCCESS;
+}
+
+static enum bt_gatt_ots_olcp_res_code olcp_next_proc_execute(
+	struct bt_gatt_ots *ots)
+{
+	int err;
+	struct bt_gatt_ots_object *next_obj;
+
+	if (!ots->cur_obj) {
+		return BT_GATT_OTS_OLCP_RES_OPERATION_FAILED;
+	}
+	err = bt_gatt_ots_obj_manager_next_obj_get(ots->obj_manager,
+						   ots->cur_obj,
+						   &next_obj);
+	if (err) {
+		return obj_manager_to_olcp_err_code_map(err);
+	}
+
+	ots->cur_obj = next_obj;
+
+	return BT_GATT_OTS_OLCP_RES_SUCCESS;
+}
+
+static enum bt_gatt_ots_olcp_res_code olcp_goto_proc_execute(
+	struct bt_gatt_ots *ots, u64_t id)
+{
+	int err;
+	struct bt_gatt_ots_object *id_obj;
+
+	err = bt_gatt_ots_obj_manager_obj_get(ots->obj_manager,
+					      id,
+					      &id_obj);
+	if (err) {
+		return obj_manager_to_olcp_err_code_map(err);
+	}
+
+	ots->cur_obj = id_obj;
+
+	return BT_GATT_OTS_OLCP_RES_SUCCESS;
+}
+
+static enum bt_gatt_ots_olcp_res_code olcp_proc_execute(
+	struct bt_gatt_ots *ots, struct bt_gatt_ots_olcp_proc *proc)
+{
+	LOG_DBG("Executing OLCP procedure with 0x%02X Op Code", proc->type);
+
+	switch (proc->type) {
+	case BT_GATT_OTS_OLCP_PROC_FIRST:
+		return olcp_first_proc_execute(ots);
+	case BT_GATT_OTS_OLCP_PROC_LAST:
+		return olcp_last_proc_execute(ots);
+	case BT_GATT_OTS_OLCP_PROC_PREV:
+		return olcp_prev_proc_execute(ots);
+	case BT_GATT_OTS_OLCP_PROC_NEXT:
+		return olcp_next_proc_execute(ots);
+	case BT_GATT_OTS_OLCP_PROC_GOTO:
+		return olcp_goto_proc_execute(ots, proc->goto_params.id);
+	case BT_GATT_OTS_OLCP_PROC_ORDER:
+	case BT_GATT_OTS_OLCP_PROC_REQ_NUM_OBJS:
+	case BT_GATT_OTS_OLCP_PROC_CLEAR_MARKING:
+		return BT_GATT_OTS_OLCP_RES_PROC_NOT_SUP;
+	default:
+		return BT_GATT_OTS_OLCP_RES_INVALID_PARAMETER;
+	}
+};
+
+static enum bt_gatt_ots_olcp_res_code olcp_command_decode(
+	const u8_t *buf, struct bt_gatt_ots_olcp_proc *proc)
+{
+	memset(proc, 0, sizeof(*proc));
+
+	proc->type = *buf++;
+
+	switch (proc->type) {
+	case BT_GATT_OTS_OLCP_PROC_FIRST:
+	case BT_GATT_OTS_OLCP_PROC_LAST:
+	case BT_GATT_OTS_OLCP_PROC_PREV:
+	case BT_GATT_OTS_OLCP_PROC_NEXT:
+		break;
+	case BT_GATT_OTS_OLCP_PROC_GOTO:
+	{
+		struct bt_gatt_ots_obj_id *id;
+
+		id = (struct bt_gatt_ots_obj_id *) &proc->goto_params.id;
+		memcpy(id->ba, buf, sizeof(id->ba));
+		break;
+	}
+	default:
+		LOG_WRN("OLCP unsupported procedure type");
+		return BT_GATT_OTS_OLCP_RES_PROC_NOT_SUP;
+	}
+
+	return BT_GATT_OTS_OLCP_RES_SUCCESS;
+}
+
+static bool olcp_command_len_verify(enum bt_gatt_ots_olcp_proc_type type,
+				    u16_t len)
+{
+	u16_t ref_len = sizeof(type);
```

Comment:
Be vary of getting the sizeof enums. Unless added as a compilation flag, enum sizes might vary from platform to platform, or from compiler to compiler AFAIK.

Reply:
Thanks for heads-up, I will replace it with a define constant